assembly/src/parsers/serde/deserialization.rs

```diff
@@ -1,29 +1,43 @@
 use super::{
-    ByteReader, Deserializable, Instruction, Node, OpCode, ProcedureId, SerializationError,
+    ByteReader, Deserializable, DeserializationError, Felt, Instruction, Node, OpCode, ProcedureId,
     MAX_PUSH_INPUTS,
 };
 
 // NODE DESERIALIZATION
 // ================================================================================================
 
 impl Deserializable for Node {
-    fn read_from(bytes: &mut ByteReader) -> Result<Self, SerializationError> {
-        let first_byte = bytes.peek_u8()?;
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let first_byte = source.peek_u8()?;
 
         if first_byte == OpCode::IfElse as u8 {
-            bytes.read_u8()?;
-            Ok(Node::IfElse(
-                Deserializable::read_from(bytes)?,
-                Deserializable::read_from(bytes)?,
-            ))
+            source.read_u8()?;
+
+            let if_block_len = source.read_u64()? as usize;
+            let if_block = Deserializable::read_batch_from(source, if_block_len)?;
+
+            let else_block_len = source.read_u64()? as usize;
+            let else_block = Deserializable::read_batch_from(source, else_block_len)?;
```

Comment:
I think `u64` for number of nodes here and for other statements is too much - let's use `u16` and add TODOs to enforce these restrictions via the constructor.